brx/crx/js/background.js

```diff
@@ -0,0 +1,89 @@
+function Background() {
+  this.tabInfo = {};
+  this.ports = {};
+
+  this.initPanelMessage();
+  this.initRuntimeMessage();
+  this.initWebRequest();
+}
+
+Background.prototype.sendSessionId = function(tabId) {
+  this.ports[tabId].postMessage({
+    type: "session-id",
+    sessionId: this.tabInfo[tabId].sessionId
+  });
+};
+
+Background.prototype.initPanelMessage = function() {
+  var handleMessages = function(msg) {
+    if (msg.type === "session-id") {
+      this.sendSessionId(msg.tabId);
+    }
+  };
+  handleMessages = handleMessages.bind(this);
+
+  var onConnect = function(newPort) {
+    this.ports[newPort.name] = newPort;
+    newPort.onMessage.addListener(handleMessages);
+  };
+  onConnect = onConnect.bind(this);
+
+  chrome.runtime.onConnect.addListener(onConnect);
+};
+
+Background.prototype.initRuntimeMessage = function() {
+  var handleMessages = function(req, sender, sendResponse) {
+    if (req.type === "request") {
+      var info = this.tabInfo[req.tabId]
+      var url = info.remoteHost + "/" + req.url;
+      REPLConsole.request(req.method, url, req.params, function(xhr) {
+        sendResponse({ status: xhr.status, responseText: xhr.responseText });
+      });
+      return true;
+    }
+  };
+  handleMessages = handleMessages.bind(this);
+
+  chrome.runtime.onMessage.addListener(handleMessages);
+};
+
+Background.prototype.initWebRequest = function() {
+  var reformat = function(details) {
+    var reduceFunc = function(obj, headerInfo) {
+      obj[headerInfo.name] = headerInfo.value;
+      return obj;
+    };
+
+    return {
+      method: details.method,
+      url: details.url,
+      headers: details.responseHeaders.reduce(reduceFunc, new Object)
+    };
+  };
+
+  var onResponse = function(details) {
+    var res = reformat(details);
+    var sessionId;
+    if (res.method === "GET" && (sessionId = res.headers["X-Web-Console-Session-Id"])) {
+      this.tabInfo[details.tabId] = {
+        sessionId: sessionId,
+        remoteHost: details.url.match(/([^:]+:\/\/[^\/]+)\/?/)[1]
+      };
+      this.sendSessionId(details.tabId);
+    }
+  };
+  onResponse = onResponse.bind(this);
+
+  var options = {
+    types: ["main_frame"],
+    urls: ["http://*/*", "https://*/*"]
+  };
+
+  chrome.webRequest.onHeadersReceived.addListener(
+    onResponse,
+    options,
+    ["responseHeaders"]
+  )
+};
+
+new Background;
```

Comment:
If we don't get to keep the instance around, why instantiate it in the first place? If its just for the side effect, should that be an object or a bag of functions?